File: BubblerAndroid/app/src/main/java/com/bubbler/android/data/repository/GraphRepository.kt
package com.bubbler.android.data.repository

import com.bubbler.android.core.auth.TokenStore
import com.bubbler.android.core.network.ApiClient
import com.bubbler.android.core.network.Endpoints
import com.bubbler.android.data.model.Post

/** Graph neighbor fetch (`GET /graph/posts/{id}/next`). */
open class GraphRepository(
    private val apiClient: ApiClient,
    private val tokenStore: TokenStore,
) {
    open suspend fun getNextGraphPosts(postId: String): List<Post> =
        apiClient.get(
            Endpoints.graphNextPosts(postId),
            token = tokenStore.requireAccessToken(),
        )
}

File: bubbler_app/lib/data/repositories/graph_repository.dart
import '../../core/api/api_client.dart';
import '../../core/api/api_exception.dart';
import '../../core/api/endpoints.dart';
import '../models/post.dart';

/// Graph neighbor walks (`getNextGraphPosts`).
class GraphRepository {
  GraphRepository(this._client);

  final ApiClient _client;

  /// Next neighbor posts for [postId] (`GET /graph/posts/{id}/next`).
  Future<List<Post>> getNextGraphPosts(String postId) async {
    final data = await _client.authorizedRequest(
      path: Endpoints.graphNext(postId),
    );
    return _decodePosts(data);
  }

  static List<Post> _decodePosts(dynamic data) {
    if (data is! List) {
      throw const ApiInvalidResponse();
    }
    return data
        .map((e) => Post.fromJson(_asJsonMap(e)))
        .toList();
  }

  static Map<String, dynamic> _asJsonMap(dynamic data) {
    if (data is Map<String, dynamic>) {
      return data;
    }
    if (data is Map) {
      return Map<String, dynamic>.from(data);
    }
    throw const ApiInvalidResponse();
  }
}
